src/rpc/server/scala/serv/src/main/scala/com/ctc/omega_edit/grpc/EditorService.scala

```diff
@@ -459,6 +460,36 @@ class EditorService(implicit val system: ActorSystem) extends Editor {
             SegmentResponse.of(in.sessionId, offset, ByteString.copyFrom(data))
           )
       }
+
+  def stopServer(kind: ServerControlKind): ServerControlResponse =
+    Await.ready(system.terminate(), 1.minute).value.get match {
```

Comment:
Don't await, this will block, you do not want that.

`system.terminate().map(_ => ServerControlResponse(kind, 0))(ExecutionContext.global)`

Minor gotcha is having to explicitly add the global executor, since the system.dispatcher will be destroyed by the time the mapped call runs.

I didnt test this.  Might not be exact but close.

Reply:
So would we still want something similar to set 0 on successful termination and 1 on failed termination to not cause the race condition you mentioned before? or does this snippet take that into account as well?

Reply:
The race is eliminated by using the Future result of terminate.  Several ways to use it, like the map above, or if you want the success/fail state then

```scala
system.terminate().onComplete {
        case Success(_) => ServerControlResponse(kind, 0)
        case Failure(_) => ServerControlResponse(kind, 1)
    }(ExecutionContext.global)
```

Reply:
So doing that way will throw an error right, because `onComplete` doesn't return a value because its a `Unit` method right?

Reply:
> Might not be exact but close.

```scala
  def stopServer(kind: ServerControlKind): Future[ServerControlResponse] = {
    val f = system.terminate()
    f.onComplete {
        case Success(_) => ServerControlResponse(kind, 0)
        case Failure(_) => ServerControlResponse(kind, 1)
    }(ExecutionContext.global)
    f
  }
```

Reply:
@jw3 Would something like this work?

```scala
    system
      .terminate()
      .transform {
        case Success(_) => Success(ServerControlResponse(kind, 0))
        case Failure(_) => Success(ServerControlResponse(kind, 1))
      }(scala.concurrent.ExecutionContext.global)
```

Reply:
I believe that would still cause an error as `f` would be `Fututre[akka.actor.Terminated]`. What I had posted seemed to return proper result just not sure if that's the proper way to do it.

Reply:
yes your transform example should be fine.  the point was to map your calculation to the completion of the future, which you are doing htere

Reply:
That makes sense thanks, would something like this be better since we are only change the response code?

```scala
    system
      .terminate()
      .transform {
        case Success(_) => Success(0)
        case Failure(_) => Success(1)
      }
      .map(respCode => ServerControlResponse(kind, respCode))(
        scala.concurrent.ExecutionContext.global
      )

```

Reply:
The intent of the first one is clearer imo.

Reply:
Sounds good thanks!